File: dynamicProgramming/minPathSum.py
"""
Input:
[
  [1,3,1],
  [1,5,1],
  [4,2,1]
]
Output: 7
Explanation: Because the path 1→3→1→1→1 minimizes the sum.
"""
class Solution(object):
    def minPathSum(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        m = len(grid)
        n = len(grid[0])
        
        """
        初始化第一行数据 
        """
        for i in range(1, n):
            grid[0][i] += grid[0][i - 1]

        """
        初始化第一列数据
        """
        for j in range(1, m):
            grid[j][0] += grid[j - 1][0]
            
        for i in range(1, m):
            for j in range(1, n):
                grid[i][j] += min(grid[i-1][j], grid[i][j-1])
        return grid[-1][-1]

File: minPathSum/minPathSum.java
class Solution {
    public int minPathSum(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int[][] states = new int[m][n];
        
        int sum = 0;
        for (int j = 0; j < n; ++j) { // 初始化 states 的第一行数据
            sum += grid[0][j];
            states[0][j] = sum;
        }
        
        sum = 0;
        for (int i = 0; i < m; ++i) { // 初始化 states 的第一列数据
            sum += grid[i][0];
            states[i][0] = sum;
        }
        
        for (int i = 1; i < m; ++i) {
            for (int j = 1; j < n; ++j) {
                states[i][j] = grid[i][j] + Math.min(states[i][j-1], states[i-1][j]);
            }
        }
        
        return states[m-1][n-1];
    }
}